# NovaGen Research Labs¶

In [29]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, recall_score,f1_score

from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression


from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import VotingClassifier

In [2]:
df=pd.read_csv("novagen_dataset.csv")
df

,Age,BMI,Blood_Pressure,Cholesterol,Glucose_Level,Heart_Rate,Sleep_Hours,Exercise_Hours,Water_Intake,Stress_Level,...,Diet,MentalHealth,PhysicalActivity,MedicalHistory,Allergies,Diet_Type__Vegan,Diet_Type__Vegetarian,Blood_Group_AB,Blood_Group_B,Blood_Group_O
0,2.0,26.0,111.0,198.0,99.0,72.0,4.0,1.0,5.0,5.0,...,1,2,1,0,1,False,True,True,False,False
1,8.0,24.0,121.0,199.0,103.0,75.0,2.0,1.0,2.0,9.0,...,1,2,1,2,2,False,False,True,False,False
2,81.0,27.0,147.0,203.0,100.0,74.0,10.0,-0.0,5.0,1.0,...,2,0,0,1,0,True,False,False,False,False
3,25.0,21.0,150.0,199.0,102.0,70.0,7.0,3.0,3.0,3.0,...,1,2,1,2,0,True,False,False,True,False
4,24.0,26.0,146.0,202.0,99.0,76.0,10.0,2.0,5.0,1.0,...,2,0,2,0,2,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9544,5.0,22.0,109.0,203.0,98.0,75.0,8.0,1.0,6.0,0.0,...,0,2,2,1,0,True,False,True,False,False
9545,94.0,26.0,144.0,203.0,96.0,72.0,8.0,4.0,2.0,1.0,...,1,0,1,0,2,False,True,False,True,False
9546,10.0,23.0,185.0,198.0,103.0,72.0,4.0,5.0,5.0,6.0,...,1,0,2,0,1,True,False,True,False,False
9547,50.0,29.0,166.0,200.0,100.0,74.0,8.0,2.0,3.0,4.0,...,2,0,0,1,1,True,False,True,False,False


In [3]:
df.describe()

,Age,BMI,Blood_Pressure,Cholesterol,Glucose_Level,Heart_Rate,Sleep_Hours,Exercise_Hours,Water_Intake,Stress_Level,Target,Smoking,Alcohol,Diet,MentalHealth,PhysicalActivity,MedicalHistory,Allergies
count,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000,9549.000000
mean,33.806786,25.660697,130.382658,199.091528,100.225678,73.613782,6.951409,1.892345,3.580899,4.382134,0.521416,0.990470,0.995183,1.005864,0.998429,1.003351,1.004713,0.989318
std,24.566473,1.942369,27.878476,1.969234,2.157999,1.681538,2.352152,1.378714,1.622874,2.078593,0.499567,0.815521,0.816653,0.815877,0.821844,0.808800,0.813506,0.815699
min,0.000000,19.000000,22.000000,192.000000,93.000000,67.000000,0.000000,-0.000000,-0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,14.000000,24.000000,113.000000,198.000000,99.000000,73.000000,5.000000,1.000000,2.000000,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,29.000000,26.000000,134.000000,199.000000,100.000000,74.000000,7.000000,2.000000,4.000000,4.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000
75%,50.000000,27.000000,150.000000,200.000000,102.000000,75.000000,9.000000,3.000000,5.000000,6.000000,1.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000
max,100.000000,32.000000,225.000000,207.000000,107.000000,80.000000,14.000000,8.000000,10.000000,12.000000,1.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9549 entries, 0 to 9548
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Age                    9549 non-null   float64
 1   BMI                    9549 non-null   float64
 2   Blood_Pressure         9549 non-null   float64
 3   Cholesterol            9549 non-null   float64
 4   Glucose_Level          9549 non-null   float64
 5   Heart_Rate             9549 non-null   float64
 6   Sleep_Hours            9549 non-null   float64
 7   Exercise_Hours         9549 non-null   float64
 8   Water_Intake           9549 non-null   float64
 9   Stress_Level           9549 non-null   float64
 10  Target                 9549 non-null   int64  
 11  Smoking                9549 non-null   int64  
 12  Alcohol                9549 non-null   int64  
 13  Diet                   9549 non-null   int64  
 14  MentalHealth           9549 non-null   int64  
 15  Phys

In [5]:
x=df.drop("Target",axis=1)
y=df["Target"]


x_train,x_test,y_train,y_test=train_test_split(
    x,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y     #both sets have same proportion of target variable values
)

# Scale values


In [6]:
from sklearn.preprocessing import StandardScaler

scaler=StandardScaler()
x_train_scaled=scaler.fit_transform(x_train)
x_test_scaled=scaler.transform(x_test)

# create your baseline model

### logistic regression with regularization

In [7]:

lr=LogisticRegression(
    penalty="l2",
    solver="liblinear",
    max_iter=1000
)
lr.fit(x_train_scaled,y_train)
y_pred=lr.predict(x_test_scaled)

# In Model Evaluation, Recall is more important than accuracy 
# because missing a high-risk patient is dangerous

print("Logistic Regression Accuracy:", accuracy_score(y_test, y_pred))
print("Logistic Regression Recall:", recall_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Logistic Regression Accuracy: 0.8141361256544503
Logistic Regression Recall: 0.8283132530120482
              precision    recall  f1-score   support

           0       0.81      0.80      0.80       914
           1       0.82      0.83      0.82       996

    accuracy                           0.81      1910
   macro avg       0.81      0.81      0.81      1910
weighted avg       0.81      0.81      0.81      1910



# Model 2 - KNN

In [10]:
knn=KNeighborsClassifier(
    n_neighbors=5,
    metric="euclidean",
    n_jobs=-1
)


knn.fit(x_train_scaled,y_train)
y_pred=knn.predict(x_test_scaled)

print("KNN  Accuracy is: ", accuracy_score(y_test,y_pred)*100,"%")
print("KNN recall score is: ", recall_score(y_test,y_pred)*100,"%")
print("KNN clasification report \n: ", classification_report(y_test,y_pred))

KNN  Accuracy is:  88.32460732984293 %
KNN recall score is:  88.35341365461848 %
KNN clasification report 
:                precision    recall  f1-score   support

           0       0.87      0.88      0.88       914
           1       0.89      0.88      0.89       996

    accuracy                           0.88      1910
   macro avg       0.88      0.88      0.88      1910
weighted avg       0.88      0.88      0.88      1910



## Random Forest (Ensemble Learning)

In [30]:
rf=RandomForestClassifier(
    n_estimators=200,
    max_depth=None,
    random_state=42
)

rf.fit(x_train_scaled,y_train)
y_pred=rf.predict(x_test_scaled)

print("RandomForest Classifier accuracy is: ", accuracy_score(y_test,y_pred)*100,"%")
print("RandomForest Classifier recall is: ", recall_score(y_test,y_pred)*100,"%")

print(classification_report(y_test,y_pred))

RandomForest Classifier accuracy is:  93.97905759162303 %
RandomForest Classifier recall is:  96.18473895582329 %
RandomForest Classifier recall is:  94.33776464795667 %
              precision    recall  f1-score   support

           0       0.96      0.92      0.94       914
           1       0.93      0.96      0.94       996

    accuracy                           0.94      1910
   macro avg       0.94      0.94      0.94      1910
weighted avg       0.94      0.94      0.94      1910



## Gradient Boosting (Ensemble Learning)

In [20]:
gb=GradientBoostingClassifier(
    n_estimators=200,
    learning_rate=0.1,
    max_depth=3,
    random_state=42
)

gb.fit(x_train_scaled,y_train)
y_pred=gb.predict(x_test_scaled)

print("Gradient Boosting Accuracy: ", accuracy_score(y_test,y_pred)*100,"%")
print("Gradient Boosting recall is: ", recall_score(y_test,y_pred)*100,"%")
print(classification_report(y_test,y_pred))



Gradient Boosting Accuracy:  93.50785340314135 %
Gradient Boosting recall is:  95.28112449799197 %
              precision    recall  f1-score   support

           0       0.95      0.92      0.93       914
           1       0.92      0.95      0.94       996

    accuracy                           0.94      1910
   macro avg       0.94      0.93      0.93      1910
weighted avg       0.94      0.94      0.94      1910



## Voting Classifier (Ensemble Learning)

In [26]:
voting_cl = VotingClassifier(
    estimators=[
        ("lr", LogisticRegression(
            max_iter=1000,
            solver="liblinear"
        )),
        
        ("knn", KNeighborsClassifier(
            n_neighbors=5
        )),
        
        ("rf", RandomForestClassifier(
            n_estimators=200,
            random_state=42
        ))
    ],

    voting="soft"
)

voting_cl.fit(x_train_scaled, y_train)

y_pred = voting_cl.predict(x_test_scaled)

print("Voting Classifier Accuracy:",
      accuracy_score(y_test, y_pred) * 100, "%")

print("Voting Classifier Recall:",
      recall_score(y_test, y_pred) * 100, "%")

print(classification_report(y_test, y_pred))

Voting Classifier Accuracy: 91.57068062827224 %
Voting Classifier Recall: 92.9718875502008 %
              precision    recall  f1-score   support

           0       0.92      0.90      0.91       914
           1       0.91      0.93      0.92       996

    accuracy                           0.92      1910
   macro avg       0.92      0.92      0.92      1910
weighted avg       0.92      0.92      0.92      1910



## Results

| Model                | Recall |
|----------------------|:------:|
| Logistic Regression  | 82.8%  |
| KNN                  | 88.3%  |
| Random Forest        | 96.1%  |
| Gradient Boosting    | 95.9%  |
| Voting Classifier    | 92.07% |

### Best Classifier that we should use for NovaGen(based on Recall) - Random Forest with accuracy nearly = 94.7%